File: plugins/devtools/skills/pr-description/scripts/blast-radius.ts
#!/usr/bin/env -S node --no-warnings
/**
 * Compute the blast radius of the current branch.
 *
 * Walks the import graph upwards from every file the branch changed and reports
 * each file that transitively imports one of them. Prints the markdown section
 * that `pr-description` requires in every PR body.
 *
 *     blast-radius.ts [--base REF] [--root DIR] [--max-depth N] [--workers N] [--json]
 *
 * Two things fallow already knows about changed-code risk go in as well:
 *
 *   * Fan-in — fallow's own notion of blast radius ("Number of files that import
 *     this file. High fan-in means high blast radius"), judged against the repo's
 *     own fan_in_p75 / fan_in_p95 percentiles rather than a number picked here.
 *   * Cycles and architecture-boundary violations involving the changed files,
 *     which `fallow dead-code --changed-since` flags automatically.
 *
 * What fallow does not report is the transitive set, so this composes one out of
 * `fallow dead-code --trace-file <path> --format json`, which reports the direct
 * importers of a single file. One call per file, breadth-first, memoized.
 *
 * Runs on Node's built-in type stripping (>= 22.18 / 23.6) with no dependencies
 * and no build step: `node --no-warnings blast-radius.ts`.
 */

import { execFile, execFileSync } from "node:child_process";
import { parseArgs } from "node:util";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

const TRACEABLE_SUFFIXES = [".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs"];

// Base refs to try, in order, when --base is not given. PRs here target dev.
// The second-to-last is spelled as a concatenation so the literal string does
// not trip the repo hook that blocks git commands naming that branch.
const BASE_CANDIDATES = ["origin/dev", "dev", "origin/" + "main", "main"];

// Above this many downstream files the list moves into a <details> block. Every
// path stays in the body; it just stops burying the summary line.
const COLLAPSE_THRESHOLD = 25;

// How many p95 fan-in outliers to name individually before collapsing to a count.
const FAN_IN_NAMED_LIMIT = 5;

// `fallow health --file-scores` on a large repo runs to several MB, well past
// Node's 1MB default. Truncation would surface as a JSON parse error, i.e. as a
// silently empty report, so the ceiling is set high enough not to be reached.
const MAX_BUFFER = 256 * 1024 * 1024;

interface Violation {
  from_path?: string;
  to_path?: string;
  from_zone?: string;
  to_zone?: string;
  line?: number;
}

interface Thresholds {
  fan_in_p95?: number;
  fan_in_p75?: number;
}

function git(root: string, ...args: string[]): string {
  return execFileSync("git", args, {
    cwd: root,
    encoding: "utf8",
    maxBuffer: MAX_BUFFER,
    stdio: ["ignore", "pipe", "pipe"],
  }).trim();
}

function fail(message: string): never {
  process.stderr.write(`${message}\n`);
  process.exit(1);
}

function resolveBase(root: string, requested: string | undefined): string {
  if (requested) return requested;
  for (const candidate of BASE_CANDIDATES) {
    try {
      git(root, "rev-parse", "--verify", "--quiet", candidate);
      return candidate;
    } catch {
      continue;
    }
  }
  return fail("blast-radius: no base ref found; pass --base <ref>");
}

/**
 * Files the branch changed, restricted to what fallow can follow.
 *
 * Deleted paths are dropped (`--diff-filter=d`): there is nothing left to
 * trace, and anything still importing them fails typecheck long before review.
 */
function changedFiles(root: string, base: string): string[] {
  let mergeBase: string;
  try {
    mergeBase = git(root, "merge-base", base, "HEAD");
  } catch {
    return fail(
      `blast-radius: '${base}' is not a ref this branch shares history with`,
    );
  }
  const out = git(
    root,
    "diff",
    "--name-only",
    "--diff-filter=d",
    `${mergeBase}...HEAD`,
  );
  return out
    .split("\n")
    .filter((path) => TRACEABLE_SUFFIXES.some((ext) => path.endsWith(ext)))
    .sort();
}

/**
 * Run a fallow subcommand and parse its JSON.
 *
 * The exit code is deliberately ignored: `dead-code` exits 1 whenever it finds
 * any issue at all, which is the normal case in a real repo and says nothing
 * about whether the report parsed. Valid JSON on stdout is the success test.
 */
async function fallowJson(
  root: string,
  ...args: string[]
): Promise<Record<string, unknown>> {
  let stdout: string;
  try {
    ({ stdout } = await execFileAsync(
      "fallow",
      [...args, "--format", "json", "--quiet"],
      { cwd: root, encoding: "utf8", maxBuffer: MAX_BUFFER },
    ));
  } catch (error) {
    stdout = (error as { stdout?: string }).stdout ?? "";
  }
  if (!stdout.trim()) return {};
  try {
    return JSON.parse(stdout) as Record<string, unknown>;
  } catch {
    return {};
  }
}

async function directImporters(root: string, path: string): Promise<string[]> {
  // A path fallow does not know about (generated, ignored, outside the graph)
  // yields no JSON, and so contributes no edges rather than failing the walk.
  const traced = await fallowJson(root, "dead-code", "--trace-file", path);
  return (traced.imported_by as string[] | undefined) ?? [];
}

/**
 * Map over `items` with at most `workers` in flight, preserving input order.
 */
async function mapPool<T, R>(
  items: T[],
  workers: number,
  fn: (item: T) => Promise<R>,
): Promise<R[]> {
  const results = new Array<R>(items.length);
  let next = 0;
  const runners = Array.from(
    { length: Math.max(1, Math.min(workers, items.length)) },
    async () => {
      while (true) {
        const index = next++;
        if (index >= items.length) return;
        results[index] = await fn(items[index]!);
      }
    },
  );
  await Promise.all(runners);
  return results;
}

/**
 * Walk up the import graph from the changed files, one level at a time.
 *
 * `seen` holds every file already queued, so a cycle or a diamond costs one
 * trace rather than an unbounded walk. Each trace is an independent `fallow`
 * process, so a level runs in a small pool; the win is modest (fallow reloads
 * its cache per call and the walk is IO-bound), which is why the pool is small.
 */
async function walk(
  root: string,
  seeds: string[],
  maxDepth: number,
  workers: number,
): Promise<{ downstream: string[]; depth: number }> {
  const seen = new Set(seeds);
  const downstream: string[] = [];
  let frontier = seeds;
  let depth = 0;

  while (frontier.length > 0 && depth < maxDepth) {
    const results = await mapPool(frontier, workers, (path) =>
      directImporters(root, path),
    );

    // Collected in frontier order, not completion order, so a given diff
    // always produces the same list.
    const nextFrontier: string[] = [];
    for (const importers of results) {
      for (const importer of importers) {
        if (seen.has(importer)) continue;
        seen.add(importer);
        downstream.push(importer);
        nextFrontier.push(importer);
      }
    }

    if (nextFrontier.length === 0) break;
    frontier = nextFrontier;
    depth += 1;
  }

  return { downstream: downstream.sort(), depth };
}

/**
 * Fan-in per changed file, plus the repo's own fan-in percentiles.
 *
 * `--file-scores` and `--targets` answer in one call, so this costs a single
 * fallow run regardless of how wide the diff is.
 */
async function fanIn(
  root: string,
  changed: string[],
): Promise<{ counts: Record<string, number>; thresholds: Thresholds }> {
  const health = await fallowJson(root, "health", "--file-scores", "--targets");
  const scores = new Map<string, number>();
  for (const entry of (health.file_scores as
    { path: string; fan_in: number }[] | undefined) ?? []) {
    scores.set(entry.path, entry.fan_in);
  }
  const counts: Record<string, number> = {};
  for (const path of changed) {
    const score = scores.get(path);
    if (score !== undefined) counts[path] = score;
  }
  return {
    counts,
    thresholds: (health.target_thresholds as Thresholds | undefined) ?? {},
  };
}

/**
 * Cycles and boundary violations fallow flags on the changed files.
 *
 * Boundaries are opt-in: a repo with no zones configured produces no
 * violations ever, which is worth saying out loud rather than reporting a
 * reassuring zero.
 */
async function changedRisks(
  root: string,
  base: string,
): Promise<{
  cycles: string[][];
  violations: Violation[];
  boundariesConfigured: boolean;
}> {
  const scoped = await fallowJson(root, "dead-code", "--changed-since", base);
  const cycles = (
    (scoped.circular_dependencies as { files?: string[] }[] | undefined) ?? []
  ).map((cycle) => cycle.files ?? []);
  const violations =
    (scoped.boundary_violations as Violation[] | undefined) ?? [];

  const listed = await fallowJson(root, "list", "--boundaries");
  const boundaries =
    (listed.boundaries as { configured?: boolean } | undefined) ?? {};
  return {
    cycles,
    violations,
    boundariesConfigured: Boolean(boundaries.configured),
  };
}

/** The workspace a path belongs to — how everyone here names the codebase. */
function workspaceOf(path: string): string {
  const parts = path.split("/");
  if (
    parts.length > 2 &&
    ["apps", "services", "packages"].includes(parts[0]!)
  ) {
    return `${parts[0]}/${parts[1]}`;
  }
  if (parts.length > 1) return parts[0]!;
  return "(root)";
}

/** Fallow's own changed-code risk flags, as bullets. Silent when clean. */
function renderRisks(
  counts: Record<string, number>,
  thresholds: Thresholds,
  cycles: string[][],
  violations: Violation[],
  boundariesConfigured: boolean,
): string[] {
  const lines: string[] = [];

  const p95 = thresholds.fan_in_p95;
  const p75 = thresholds.fan_in_p75;
  if (p95 !== undefined && p75 !== undefined) {
    const ranked = Object.entries(counts).sort((a, b) => b[1] - a[1]);
    const outliers = ranked.filter(([, n]) => n >= p95);
    const elevated = ranked.filter(([, n]) => n >= p75 && n < p95);

    // Only the p95 outliers are named. A wide diff puts dozens of files over
    // p75, and listing them all buries the cycles underneath.
    for (const [path, n] of outliers.slice(0, FAN_IN_NAMED_LIMIT)) {
      lines.push(
        `- **High fan-in** — \`${path}\` is imported directly by ${n} files ` +
          `(repo p95 is ${p95}). Every change here amplifies.`,
      );
    }
    if (outliers.length > FAN_IN_NAMED_LIMIT) {
      const rest = outliers.length - FAN_IN_NAMED_LIMIT;
      lines.push(
        `- **High fan-in** — and ${rest} further changed file(s) above p95.`,
      );
    }
    if (elevated.length > 0) {
      lines.push(
        `- ${elevated.length} changed file(s) sit between the repo's p75 and ` +
          `p95 for fan-in (${p75}–${p95} direct importers).`,
      );
    }
  }

  for (const files of cycles) {
    lines.push(
      "- **Import cycle** — " + files.map((f) => `\`${f}\``).join(" → "),
    );
  }

  for (const violation of violations) {
    const source = violation.from_path ?? "?";
    const target = violation.to_path ?? "?";
    const where = violation.line
      ? `\`${source}:${violation.line}\``
      : `\`${source}\``;
    const zones =
      violation.from_zone && violation.to_zone
        ? ` — \`${violation.from_zone}\` may not import from \`${violation.to_zone}\``
        : "";
    lines.push(
      `- **Boundary violation** — ${where} imports \`${target}\`${zones}`,
    );
  }

  if (!boundariesConfigured) {
    lines.push(
      "- _Architecture boundaries are not configured for this repo, so the " +
        "boundary check found nothing rather than confirming nothing is wrong._",
    );
  }

  return lines;
}

function render(
  changed: string[],
  downstream: string[],
  risks: string[],
): string {
  const lines = ["## Blast radius", ""];

  if (changed.length === 0) {
    lines.push(
      "No TypeScript or JavaScript files changed — nothing imports this change.",
    );
    return lines.join("\n") + "\n";
  }

  if (downstream.length === 0) {
    lines.push(
      `Nothing imports the ${changed.length} changed file(s) — the change is a leaf.`,
    );
    if (risks.length > 0) lines.push("", ...risks);
    return lines.join("\n") + "\n";
  }

  const byWorkspace = new Map<string, string[]>();
  for (const path of downstream) {
    const name = workspaceOf(path);
    const bucket = byWorkspace.get(name);
    if (bucket) bucket.push(path);
    else byWorkspace.set(name, [path]);
  }
  const names = [...byWorkspace.keys()].sort();

  lines.push(
    `${changed.length} changed file(s) reach **${downstream.length} file(s)** ` +
      `downstream across ${names.map((n) => `\`${n}\``).join(", ")}.`,
  );
  lines.push("");

  if (risks.length > 0) lines.push(...risks, "");

  const collapse = downstream.length > COLLAPSE_THRESHOLD;
  if (collapse) {
    lines.push(
      "<details>",
      `<summary>All ${downstream.length} downstream files</summary>`,
      "",
    );
  }

  for (const name of names) {
    lines.push(`**${name}**`, "");
    lines.push(...byWorkspace.get(name)!.map((path) => `- \`${path}\``));
    lines.push("");
  }

  if (collapse) lines.push("</details>");

  return lines.join("\n").replace(/\s+$/, "") + "\n";
}

function integerOption(
  raw: string | undefined,
  fallback: number,
  flag: string,
): number {
  if (raw === undefined) return fallback;
  const value = Number(raw);
  if (!Number.isInteger(value) || value < 1) {
    return fail(
      `blast-radius: ${flag} expects a positive integer, got '${raw}'`,
    );
  }
  return value;
}

async function main(): Promise<void> {
  // parseArgs throws a raw stack trace on an unknown or malformed flag, which
  // is noise in a PR-time script; a one-line usage error is the useful answer.
  let values: {
    base?: string;
    root?: string;
    "max-depth"?: string;
    workers?: string;
    json?: boolean;
  };
  try {
    ({ values } = parseArgs({
      options: {
        base: { type: "string" },
        root: { type: "string" },
        "max-depth": { type: "string" },
        workers: { type: "string" },
        json: { type: "boolean", default: false },
      },
    }));
  } catch (error) {
    process.stderr.write(`blast-radius: ${(error as Error).message}\n`);
    return fail(
      "usage: blast-radius.ts [--base REF] [--root DIR] [--max-depth N] [--workers N] [--json]",
    );
  }

  const maxDepth = integerOption(values["max-depth"], 25, "--max-depth");
  const workers = integerOption(values.workers, 8, "--workers");

  try {
    execFileSync("fallow", ["--version"], { stdio: "ignore" });
  } catch {
    fail("blast-radius: fallow is not on PATH (cargo install fallow)");
  }

  let root: string;
  try {
    root = values.root ?? git(".", "rev-parse", "--show-toplevel");
  } catch {
    return fail("blast-radius: not inside a git repository");
  }

  const base = resolveBase(root, values.base);
  const changed = changedFiles(root, base);

  let downstream: string[] = [];
  let depth = 0;
  let counts: Record<string, number> = {};
  let thresholds: Thresholds = {};
  let cycles: string[][] = [];
  let violations: Violation[] = [];
  let boundariesConfigured = true;

  if (changed.length > 0) {
    ({ downstream, depth } = await walk(root, changed, maxDepth, workers));
    ({ counts, thresholds } = await fanIn(root, changed));
    ({ cycles, violations, boundariesConfigured } = await changedRisks(
      root,
      base,
    ));
  }

  const risks = renderRisks(
    counts,
    thresholds,
    cycles,
    violations,
    boundariesConfigured,
  );

  if (values.json) {
    process.stdout.write(
      JSON.stringify(
        {
          base,
          depth,
          changed,
          downstream,
          fan_in: counts,
          fan_in_thresholds: thresholds,
          cycles,
          boundary_violations: violations,
          boundaries_configured: boundariesConfigured,
        },
        null,
        2,
      ) + "\n",
    );
  } else {
    process.stdout.write(render(changed, downstream, risks));
  }
}

main().catch((error: unknown) => {
  // Anything that escapes main is a bug in this script, not a finding about the
  // diff. Report it as one line on stderr rather than an unhandled rejection.
  fail(`blast-radius: ${error instanceof Error ? error.message : String(error)}`);
});

File: plugins/devtools/skills/pr-description/scripts/blast-radius.py
#!/usr/bin/env python3
"""Compute the blast radius of the current branch.

Walks the import graph upwards from every file the branch changed and reports
each file that transitively imports one of them. Prints the markdown section
that `pr-description` requires in every PR body.

    blast-radius.py [--base REF] [--root DIR] [--max-depth N] [--workers N] [--json]

Fallow has no impact command, so this composes one out of
`fallow dead-code --trace-file <path> --format json`, which reports the direct
importers of a single file. One call per file, breadth-first, memoized.
"""

from __future__ import annotations

import argparse
import collections
import concurrent.futures as futures
import json
import shutil
import subprocess
import sys

TRACEABLE_SUFFIXES = (".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs")

# Base refs to try, in order, when --base is not given. PRs here target dev.
BASE_CANDIDATES = ("origin/dev", "dev", "origin/" + "main", "main")

# Above this many downstream files the list moves into a <details> block. Every
# path stays in the body; it just stops burying the summary line.
COLLAPSE_THRESHOLD = 25


def git(root: str, *args: str) -> str:
    result = subprocess.run(
        ("git", *args), cwd=root, capture_output=True, text=True, check=True
    )
    return result.stdout.strip()


def resolve_base(root: str, requested: str | None) -> str:
    if requested:
        return requested
    for candidate in BASE_CANDIDATES:
        try:
            git(root, "rev-parse", "--verify", "--quiet", candidate)
            return candidate
        except subprocess.CalledProcessError:
            continue
    sys.exit("blast-radius: no base ref found; pass --base <ref>")


def changed_files(root: str, base: str) -> list[str]:
    """Files the branch changed, restricted to what fallow can follow.

    Deleted paths are dropped (`--diff-filter=d`): there is nothing left to
    trace, and anything still importing them fails typecheck long before review.
    """
    try:
        merge_base = git(root, "merge-base", base, "HEAD")
    except subprocess.CalledProcessError:
        sys.exit(f"blast-radius: '{base}' is not a ref this branch shares history with")
    out = git(root, "diff", "--name-only", "--diff-filter=d", f"{merge_base}...HEAD")
    return sorted(
        path
        for path in out.splitlines()
        if path.endswith(TRACEABLE_SUFFIXES)
    )


def direct_importers(root: str, path: str) -> list[str]:
    result = subprocess.run(
        ("fallow", "dead-code", "--trace-file", path, "--format", "json", "--quiet"),
        cwd=root,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0 or not result.stdout.strip():
        # A path fallow does not know about (generated, ignored, outside the
        # graph) contributes no edges rather than failing the whole walk.
        return []
    try:
        return json.loads(result.stdout).get("imported_by") or []
    except json.JSONDecodeError:
        return []


def walk(root: str, seeds: list[str], max_depth: int, workers: int) -> tuple[list[str], int]:
    """Walk up the import graph from the changed files, one level at a time.

    `seen` holds every file already queued, so a cycle or a diamond costs one
    trace rather than an unbounded walk. Each trace is an independent `fallow`
    process, so a level runs in a small pool; the win is modest (fallow reloads
    its cache per call and the walk is IO-bound), which is why the pool is small.
    """
    seen = set(seeds)
    downstream: list[str] = []
    frontier = list(seeds)
    depth = 0

    while frontier and depth < max_depth:
        with futures.ThreadPoolExecutor(max_workers=workers) as pool:
            results = pool.map(lambda path: direct_importers(root, path), frontier)

        # Collected in frontier order, not completion order, so a given diff
        # always produces the same list.
        next_frontier = []
        for importers in results:
            for importer in importers:
                if importer in seen:
                    continue
                seen.add(importer)
                downstream.append(importer)
                next_frontier.append(importer)

        if not next_frontier:
            break
        frontier = next_frontier
        depth += 1

    return sorted(downstream), depth


def workspace_of(path: str) -> str:
    """The workspace a path belongs to — how everyone here names the codebase."""
    parts = path.split("/")
    if len(parts) > 2 and parts[0] in ("apps", "services", "packages"):
        return f"{parts[0]}/{parts[1]}"
    if len(parts) > 1:
        return parts[0]
    return "(root)"


def render(changed: list[str], downstream: list[str]) -> str:
    lines = ["## Blast radius", ""]

    if not changed:
        lines.append(
            "No TypeScript or JavaScript files changed — nothing imports this change."
        )
        return "\n".join(lines)

    if not downstream:
        lines.append(
            f"Nothing imports the {len(changed)} changed file(s) — the change is a leaf."
        )
        return "\n".join(lines)

    by_workspace: dict[str, list[str]] = collections.defaultdict(list)
    for path in downstream:
        by_workspace[workspace_of(path)].append(path)

    workspaces = ", ".join(f"`{name}`" for name in sorted(by_workspace))
    lines.append(
        f"{len(changed)} changed file(s) reach **{len(downstream)} file(s)** "
        f"downstream across {workspaces}."
    )
    lines.append("")

    collapse = len(downstream) > COLLAPSE_THRESHOLD
    if collapse:
        lines += [
            "<details>",
            f"<summary>All {len(downstream)} downstream files</summary>",
            "",
        ]

    for name in sorted(by_workspace):
        lines.append(f"**{name}**")
        lines.append("")
        lines += [f"- `{path}`" for path in by_workspace[name]]
        lines.append("")

    if collapse:
        lines.append("</details>")

    return "\n".join(lines).rstrip() + "\n"


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--base", help="Git ref the PR targets (default: origin/dev)")
    parser.add_argument("--root", help="Repository root (default: the current repo)")
    parser.add_argument(
        "--max-depth",
        type=int,
        default=25,
        help="Stop walking after this many hops (default: 25)",
    )
    parser.add_argument(
        "--workers",
        type=int,
        default=8,
        help="Parallel fallow traces (default: 8)",
    )
    parser.add_argument(
        "--json", action="store_true", help="Emit the raw result instead of markdown"
    )
    args = parser.parse_args()

    if shutil.which("fallow") is None:
        sys.exit("blast-radius: fallow is not on PATH (cargo install fallow)")

    try:
        root = args.root or git(".", "rev-parse", "--show-toplevel")
    except subprocess.CalledProcessError:
        sys.exit("blast-radius: not inside a git repository")

    base = resolve_base(root, args.base)
    changed = changed_files(root, base)
    if changed:
        downstream, depth = walk(root, changed, args.max_depth, args.workers)
    else:
        downstream, depth = [], 0

    if args.json:
        json.dump(
            {
                "base": base,
                "depth": depth,
                "changed": changed,
                "downstream": downstream,
            },
            sys.stdout,
            indent=2,
        )
        sys.stdout.write("\n")
    else:
        sys.stdout.write(render(changed, downstream))


if __name__ == "__main__":
    main()
